File: iOS/CoralReefPlayer/CoralReefPlayer.swift
import Foundation
import CoralReefPlayerIOS

public enum Transport : Int32 {
    case UDP = 0
    case TCP = 1
}

public enum VideoFormat : Int32 {
    case YUV420P = 0
    case NV12 = 1
    case NV21 = 2
    case RGB24 = 3
    case BGR24 = 4
    case ARGB32 = 5
    case RGBA32 = 6
    case ABGR32 = 7
    case BGRA32 = 8
}

public enum AudioFormat : Int32 {
    case U8 = 0
    case S16 = 1
    case S32 = 2
    case F32 = 3
}

public enum Event : Int32 {
    case NEW_FRAME = 0
    case ERROR = 1
    case START = 2
    case PLAYING = 3
    case END = 4
    case STOP = 5
    case NEW_AUDIO = 6
}

public struct Option {
    public var transport: Transport
    public var width: Int
    public var height: Int
    public var videoFormat: VideoFormat
    public var hwDevice: String
    public var enableAudio: Bool
    public var sampleRate: Int
    public var channels: Int
    public var audioFormat: AudioFormat
    public var timeout: Int64
    
    public init(transport: Transport = .UDP, width: Int = 0, height: Int = 0, videoFormat: VideoFormat = .YUV420P, hwDevice: String = "",
        enableAudio: Bool = false, sampleRate: Int = 0, channels: Int = 0, audioFormat: AudioFormat = .U8, timeout: Int64 = 0) {
        self.transport = transport
        self.width = width
        self.height = height
        self.videoFormat = videoFormat
        self.hwDevice = hwDevice
        self.enableAudio = enableAudio
        self.sampleRate = sampleRate
        self.channels = channels
        self.audioFormat = audioFormat
        self.timeout = timeout
    }
}

public struct Frame {
    public var width: Int = 0
    public var height: Int = 0
    public var sampleRate: Int = 0
    public var channels: Int = 0
    public var format: Int32
    public var data: [UnsafeBufferPointer<UInt8>]
    public var linesize: [Int]
    public var pts: UInt64
}

private func swift_callback(event: Int32, data: UnsafeMutableRawPointer?, userData: UnsafeMutableRawPointer?) {
    let player = Unmanaged<Player>.fromOpaque(userData!).takeUnretainedValue()
    let ee = Event(rawValue: event)!
    if ee == Event.NEW_FRAME {
        let cframe = data!.bindMemory(to: CoralReefPlayerIOS.Frame.self, capacity: 1).pointee
        let frame = Frame(
            width: Int(cframe.width),
            height: Int(cframe.height),
            format: cframe.format,
            data: [
                UnsafeBufferPointer(start: cframe.data.0, count: Int(cframe.linesize.0 * cframe.height)),
                UnsafeBufferPointer(start: cframe.data.1, count: Int(cframe.linesize.1 * cframe.height / 2)),
                UnsafeBufferPointer(start: cframe.data.2, count: Int(cframe.linesize.2 * cframe.height / 2)),
                UnsafeBufferPointer(start: cframe.data.3, count: Int(cframe.linesize.3 * cframe.height / 2))
            ],
            linesize: [Int(cframe.linesize.0), Int(cframe.linesize.1), Int(cframe.linesize.2), Int(cframe.linesize.3)],
            pts: cframe.pts
        )
        player.callback(ee, frame)
    } else if ee == Event.NEW_AUDIO {
        let cframe = data!.bindMemory(to: CoralReefPlayerIOS.Frame.self, capacity: 1).pointee
        let frame = Frame(
            width: Int(cframe.width),
            height: Int(cframe.height),
            format: cframe.format,
            data: [
                UnsafeBufferPointer(start: cframe.data.0, count: Int(cframe.linesize.0)),
            ],
            linesize: [Int(cframe.linesize.0)],
            pts: cframe.pts
        )
        player.callback(ee, frame)
    } else {
        player.callback(ee, data == nil ? 0 : Int(bitPattern: data!))
    }
}

public class Player {

    var handle: crp_handle?
    var callback: (Event, Any) -> Void = { _, _ in }


    public init() {
        handle = crp_create()
    }

    deinit {
        if let handle = handle {
            crp_destroy(handle)
        }
    }

    public func auth(username: String, password: String, isMd5: Bool) {
        crp_auth(handle, username, password, isMd5)
    }

    public func play(url: String, option: Option, callback: @escaping (Event, Any) -> Void) {
        var opt = CoralReefPlayerIOS.Option(
            transport: option.transport.rawValue,
            video: CoralReefPlayerIOS.Option.__Unnamed_struct_video(
                width: Int32(option.width),
                height: Int32(option.height),
                format: option.videoFormat.rawValue,
                hw_device: (0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0)
            ),
            enable_audio: option.enableAudio,
            audio: CoralReefPlayerIOS.Option.__Unnamed_struct_audio(
                sample_rate: Int32(option.sampleRate),
                channels: Int32(option.channels),
                format: option.audioFormat.rawValue
            ),
            timeout: option.timeout
        )
        option.hwDevice.withCString { hw_device in
            withUnsafeMutablePointer(to: &opt.video.hw_device) { ptr in
                strncpy(ptr, hw_device, 32)
            }
        }
        self.callback = callback
        withUnsafeMutablePointer(to: &opt) { ptr in
            crp_play(handle, url, ptr, swift_callback, Unmanaged.passUnretained(self).toOpaque())
        }
    }

    public func replay() {
        crp_replay(handle)
    }

    public func stop() {
        crp_stop(handle)
    }
}

public func versionCode() -> Int {
    return Int(crp_version_code())
}

public func versionStr() -> String {
    return String(cString: crp_version_str())
}

File: binding/csharp/CoralReefPlayer/CoralReefPlayer.cs
using System;
using System.IO;
using System.Runtime.InteropServices;
using static CoralReefPlayer.NativeMethods;

namespace CoralReefPlayer
{
    public enum Transport
    {
        UDP = 0,
        TCP = 1,
    }

    public enum Format
    {
        YUV420P = 0,
        NV12 = 1,
        NV21 = 2,
        RGB24 = 3,
        BGR24 = 4,
        ARGB32 = 5,
        RGBA32 = 6,
        ABGR32 = 7,
        BGRA32 = 8,

        U8 = 0,
        S16 = 1,
        S32 = 2,
        F32 = 3,
    }

    public enum Event
    {
        NEW_FRAME = 0,
        ERROR = 1,
        START = 2,
        PLAYING = 3,
        END = 4,
        STOP = 5,
        NEW_AUDIO = 6,
    }

    public struct Option
    {
        public Transport Transport;
        public int Width;
        public int Height;
        public Format VideoFormat;
        public string HWDevice;
        public bool EnableAudio;
        public int SampleRate;
        public int Channels;
        public Format AudioFormat;
        public long Timeout;
    }

    public struct Frame
    {
        public int Width;
        public int Height;
        public int SampleRate;
        public int Channels;
        public Format Format;
        public UnmanagedMemoryStream[] Data;
        public int[] LineSize;
        public ulong PTS;
    }

    public interface ICallback
    {
        void OnEvent(Event ev, long data);
        void OnFrame(bool isAudio, Frame frame);
    }

    public class ActionCallback : ICallback
    {
        private Action<Event, long> EventListener;
        private Action<bool, Frame> FrameListener;

        public ActionCallback(Action<Event, long> eventListener, Action<bool, Frame> frameListener)
        {
            EventListener = eventListener;
            FrameListener = frameListener;
        }

        public void OnEvent(Event ev, long data)
        {
            EventListener?.Invoke(ev, data);
        }

        public void OnFrame(bool isAudio, Frame frame)
        {
            FrameListener?.Invoke(isAudio, frame);
        }
    }

    public class Player
    {
        private IntPtr handle;
        private crp_callback cs_callback;

        public Player()
        {
            handle = crp_create();
        }

        public void Release()
        {
            crp_destroy(handle);
            handle = IntPtr.Zero;
        }

        public void Auth(string username, string password, bool isMD5)
        {
            crp_auth(handle, username, password, isMD5);
        }

        public void Play(string url, Option option, ICallback callback)
        {
            COption cOption = new COption
            {
                transport = (int)option.Transport,
                video_width = option.Width,
                video_height = option.Height,
                video_format = (int)option.VideoFormat,
                hw_device = option.HWDevice,
                enable_audio = option.EnableAudio,
                audio_sample_rate = option.SampleRate,
                audio_channels = option.Channels,
                audio_format = (int)option.AudioFormat,
                timeout = option.Timeout,
            };
            cs_callback = (ev, data, userData) =>
            {
                Event ev2 = (Event)ev;
                if (ev2 == Event.NEW_FRAME)
                {
                    CFrame cFrame = Marshal.PtrToStructure<CFrame>(data);
                    Frame frame = new Frame
                    {
                        Width = cFrame.width,
                        Height = cFrame.height,
                        Format = (Format)cFrame.format,
                        Data = new UnmanagedMemoryStream[4],
                        LineSize = new int[4],
                        PTS = cFrame.pts,
                    };
                    unsafe
                    {
                        for (int i = 0; i < 4; i++)
                        {
                            byte* pData = (byte*)cFrame.data[i].ToPointer();
                            int length = cFrame.linesize[i] * cFrame.height;
                            if (i > 0) length /= 2;
                            if (pData != null)
                                frame.Data[i] = new UnmanagedMemoryStream(pData, length);
                            frame.LineSize[i] = cFrame.linesize[i];
                        }
                    }
                    callback.OnFrame(false, frame);
                }
                else if (ev2 == Event.NEW_AUDIO)
                {
                    CFrame cFrame = Marshal.PtrToStructure<CFrame>(data);
                    Frame frame = new Frame
                    {
                        SampleRate = cFrame.width,
                        Channels = cFrame.height,
                        Format = (Format)cFrame.format,
                        Data = new UnmanagedMemoryStream[1],
                        LineSize = new int[1],
                        PTS = cFrame.pts,
                    };
                    unsafe
                    {
                        byte* pData = (byte*)cFrame.data[0].ToPointer();
                        int length = cFrame.linesize[0];
                        if (pData != null)
                            frame.Data[0] = new UnmanagedMemoryStream(pData, length);
                        frame.LineSize[0] = cFrame.linesize[0];
                    }
                    callback.OnFrame(true, frame);
                }
                else
                {
                    callback.OnEvent(ev2, data.ToInt64());
                }
            };
            crp_play(handle, url, ref cOption, cs_callback, IntPtr.Zero);
        }

        public void Replay()
        {
            crp_replay(handle);
        }

        public void Stop()
        {
            crp_stop(handle);
        }

        public static int VersionCode()
        {
            return crp_version_code();
        }

        public static string VersionStr()
        {
            IntPtr pStr = crp_version_str();
            return Marshal.PtrToStringAnsi(pStr);
        }
    }
}
